File: node/src/commands/load.ts
import fm = require("front-matter");
import { minimatch } from "minimatch";
import { readFile } from "node:fs/promises";
import { RuleCacheEntry } from "..";
import { loadConfig } from "../loadConfig";
import { cjs } from "../utils/cjs";

export async function load(agentName: string, filePath: string) {
  const config = await loadConfig();
  const agentConfig = config.agents[agentName];

  if (!agentConfig) {
    throw new Error(`Agent '${agentName}' not found in configuration.`);
  }

  // Load unified cache file
  const CACHE_FILE = ".claude/agents/code-editor/rules-cache-generated.json";
  const allAgentRules: Record<string, RuleCacheEntry[]> = JSON.parse(
    await readFile(CACHE_FILE, "utf-8")
  );

  // Collect all agents to load (current + references)
  const agentsToLoad: string[] = [];

  // Add current agent first (highest priority in topology)
  agentsToLoad.push(agentName);

  // Add referenced agents after (lower priority in topology)
  if (agentConfig.references && agentConfig.references.length > 0) {
    agentsToLoad.push(...agentConfig.references);
  }

  // Load rules from all specified agents with depth tracking
  // Lower agentDepth = higher priority (current agent has depth 0)
  const allRules: Array<RuleCacheEntry & { agentDepth: number }> = [];
  for (let i = 0; i < agentsToLoad.length; i++) {
    const agent = agentsToLoad[i];
    if (allAgentRules[agent]) {
      for (const rule of allAgentRules[agent]) {
        allRules.push({ ...rule, agentDepth: i });
      }
    } else {
      console.error(`Warning: No rules found for agent '${agent}' in cache`);
    }
  }

  // Build a tag map for quick lookup
  type RuleWithDepth = RuleCacheEntry & { agentDepth: number };
  const tagMap = new Map<string, RuleWithDepth[]>();
  for (const rule of allRules) {
    for (const tag of rule.tags) {
      if (!tagMap.has(tag)) {
        tagMap.set(tag, []);
      }
      tagMap.get(tag)!.push(rule);
    }
  }

  // Find top-level rules that match the file path
  const topLevelRules: RuleWithDepth[] = [];
  for (const rule of allRules) {
    const patterns = Array.isArray(rule.patterns)
      ? rule.patterns
      : [rule.patterns];
    const ignorePatterns = rule.ignorePatterns;

    const matchesPattern = patterns.some((pattern) =>
      minimatch(filePath, pattern)
    );
    const matchesIgnore = ignorePatterns.some((pattern) =>
      minimatch(filePath, pattern)
    );

    if (matchesPattern && !matchesIgnore) {
      topLevelRules.push(rule);
    }
  }

  // Collect all rules to load (including referenced rules)
  const rulesToLoad = new Set<RuleWithDepth>();
  const processRule = (rule: RuleWithDepth, isTopLevel: boolean) => {
    if (rulesToLoad.has(rule)) return;
    rulesToLoad.add(rule);

    // Add referencesAlways
    for (const tag of rule.referencesAlways) {
      const referencedRules = tagMap.get(tag) ?? [];
      for (const referencedRule of referencedRules) {
        processRule(referencedRule, false);
      }
    }

    // Add referencesIfTop only if this is a top-level rule
    if (isTopLevel) {
      for (const tag of rule.referencesIfTop) {
        const referencedRules = tagMap.get(tag) ?? [];
        for (const referencedRule of referencedRules) {
          processRule(referencedRule, false);
        }
      }
    }
  };

  for (const rule of topLevelRules) {
    processRule(rule, true);
  }

  if (rulesToLoad.size === 0) {
    console.log(`No additional context found for ${filePath}. Continue.`);
    return;
  }

  // Sort by [priority ASC, order DESC, agentDepth ASC, filePath ASC] for priority filtering
  // This ensures that when we filter, lower-order rules survive
  // agentDepth ASC means current agent (depth 0) rules survive over referenced agent rules
  let candidateRules = Array.from(rulesToLoad);
  candidateRules.sort((a, b) => {
    const priorityA = a.priority ?? Infinity;
    const priorityB = b.priority ?? Infinity;
    if (priorityA !== priorityB) {
      return priorityA - priorityB;
    }

    const orderA = a.order ?? Infinity;
    const orderB = b.order ?? Infinity;
    if (orderA !== orderB) {
      return orderB - orderA; // DESC: higher order gets filtered out first
    }

    // Add topology sort: lower agentDepth (current agent) survives
    if (a.agentDepth !== b.agentDepth) {
      return a.agentDepth - b.agentDepth; // ASC: current agent (0) before referenced agents (1, 2, ...)
    }

    return a.path.localeCompare(b.path);
  });

  // Apply priority filtering:
  // For each rule (in sorted order), if totalRulesToPrint > priority, skip it.
  // Each time we skip a rule, decrement totalRulesToPrint.
  const finalRules: RuleWithDepth[] = [];
  let totalRulesToPrint = candidateRules.length;

  for (const rule of candidateRules) {
    const priority = rule.priority ?? Infinity;
    if (totalRulesToPrint > priority) {
      // Skip this rule and decrement the count
      totalRulesToPrint--;
    } else {
      // Include this rule
      finalRules.push(rule);
    }
  }

  // Sort final rules by [order ASC, agentDepth ASC, filePath ASC] for output
  // agentDepth ASC means current agent rules appear first
  finalRules.sort((a, b) => {
    const orderA = a.order ?? Infinity;
    const orderB = b.order ?? Infinity;
    if (orderA !== orderB) {
      return orderA - orderB;
    }

    // Add agentDepth for output ordering
    if (a.agentDepth !== b.agentDepth) {
      return a.agentDepth - b.agentDepth;
    }

    return a.path.localeCompare(b.path);
  });

  // Print rules
  for (const rule of finalRules) {
    console.log(cjs(fm)(await readFile(rule.path, "utf-8")).body);
  }

  console.log(
    `* * *\n\nEnd of additional context for ${filePath}. Continue.`
  );
}

File: go/commands/load.go
package commands

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"sort"
	"strings"

	"github.com/bmatcuk/doublestar/v4"
	"github.com/dirt-rain/code-editor-agent/config"
	"github.com/dirt-rain/code-editor-agent/models"
)

// Load loads and prints relevant rules for a given file path
func Load(agentName, filePath string) error {
	cfg, err := config.LoadConfig()
	if err != nil {
		return err
	}

	agentConfig, ok := cfg.Agents[agentName]
	if !ok {
		return fmt.Errorf("Agent '%s' not found in configuration.", agentName)
	}

	// Load unified cache file
	cacheContent, err := os.ReadFile(models.RuleCacheFilePath)
	if err != nil {
		return fmt.Errorf("failed to read cache file: %w", err)
	}

	var allAgentRules map[string][]models.RuleCacheEntry
	if err := json.Unmarshal(cacheContent, &allAgentRules); err != nil {
		return fmt.Errorf("failed to parse cache file: %w", err)
	}

	// Collect all agents to load (current + references)
	agentsToLoad := []string{agentName}
	if agentConfig.References != nil {
		agentsToLoad = append(agentsToLoad, agentConfig.References...)
	}

	// Load rules from all specified agents with depth tracking
	allRules := []models.RuleWithDepth{}
	for i, agent := range agentsToLoad {
		if rules, ok := allAgentRules[agent]; ok {
			for _, rule := range rules {
				allRules = append(allRules, models.RuleWithDepth{
					RuleCacheEntry: rule,
					AgentDepth:     i,
				})
			}
		} else {
			fmt.Fprintf(os.Stderr, "Warning: No rules found for agent '%s' in cache\n", agent)
		}
	}

	// Build a tag map for quick lookup
	tagMap := make(map[string][]models.RuleWithDepth)
	for _, rule := range allRules {
		for _, tag := range rule.Tags {
			tagMap[tag] = append(tagMap[tag], rule)
		}
	}

	// Find top-level rules that match the file path
	topLevelRules := []models.RuleWithDepth{}
	for _, rule := range allRules {
		patterns := rule.GetPatterns()
		ignorePatterns := rule.IgnorePatterns

		matchesPattern := false
		for _, pattern := range patterns {
			matched, _ := doublestar.Match(pattern, filePath)
			if matched {
				matchesPattern = true
				break
			}
		}

		matchesIgnore := false
		for _, pattern := range ignorePatterns {
			matched, _ := doublestar.Match(pattern, filePath)
			if matched {
				matchesIgnore = true
				break
			}
		}

		if matchesPattern && !matchesIgnore {
			topLevelRules = append(topLevelRules, rule)
		}
	}

	// Collect all rules to load (including referenced rules)
	rulesToLoad := make(map[*models.RuleWithDepth]bool)
	var processRule func(rule models.RuleWithDepth, isTopLevel bool)
	processRule = func(rule models.RuleWithDepth, isTopLevel bool) {
		// Check if already processed
		for r := range rulesToLoad {
			if r.Path == rule.Path && r.AgentDepth == rule.AgentDepth {
				return
			}
		}
		ruleCopy := rule
		rulesToLoad[&ruleCopy] = true

		// Add referencesAlways
		for _, tag := range rule.ReferencesAlways {
			if referencedRules, ok := tagMap[tag]; ok {
				for _, referencedRule := range referencedRules {
					processRule(referencedRule, false)
				}
			}
		}

		// Add referencesIfTop only if this is a top-level rule
		if isTopLevel {
			for _, tag := range rule.ReferencesIfTop {
				if referencedRules, ok := tagMap[tag]; ok {
					for _, referencedRule := range referencedRules {
						processRule(referencedRule, false)
					}
				}
			}
		}
	}

	for _, rule := range topLevelRules {
		processRule(rule, true)
	}

	if len(rulesToLoad) == 0 {
		fmt.Printf("No additional context found for %s. Continue.\n", filePath)
		return nil
	}

	// Convert map to slice
	candidateRules := make([]models.RuleWithDepth, 0, len(rulesToLoad))
	for rule := range rulesToLoad {
		candidateRules = append(candidateRules, *rule)
	}

	// Sort by [priority ASC, order DESC, agentDepth ASC, filePath ASC] for priority filtering
	sort.Slice(candidateRules, func(i, j int) bool {
		priorityA := math.MaxInt32
		priorityB := math.MaxInt32
		if candidateRules[i].Priority != nil {
			priorityA = *candidateRules[i].Priority
		}
		if candidateRules[j].Priority != nil {
			priorityB = *candidateRules[j].Priority
		}
		if priorityA != priorityB {
			return priorityA < priorityB
		}

		orderA := math.MaxInt32
		orderB := math.MaxInt32
		if candidateRules[i].Order != nil {
			orderA = *candidateRules[i].Order
		}
		if candidateRules[j].Order != nil {
			orderB = *candidateRules[j].Order
		}
		if orderA != orderB {
			return orderA > orderB // DESC: higher order gets filtered out first
		}

		if candidateRules[i].AgentDepth != candidateRules[j].AgentDepth {
			return candidateRules[i].AgentDepth < candidateRules[j].AgentDepth // ASC
		}

		return candidateRules[i].Path < candidateRules[j].Path
	})

	// Apply priority filtering
	finalRules := []models.RuleWithDepth{}
	totalRulesToPrint := len(candidateRules)

	for _, rule := range candidateRules {
		priority := math.MaxInt32
		if rule.Priority != nil {
			priority = *rule.Priority
		}
		if totalRulesToPrint > priority {
			// Skip this rule and decrement the count
			totalRulesToPrint--
		} else {
			// Include this rule
			finalRules = append(finalRules, rule)
		}
	}

	// Sort final rules by [order ASC, agentDepth ASC, filePath ASC] for output
	sort.Slice(finalRules, func(i, j int) bool {
		orderA := math.MaxInt32
		orderB := math.MaxInt32
		if finalRules[i].Order != nil {
			orderA = *finalRules[i].Order
		}
		if finalRules[j].Order != nil {
			orderB = *finalRules[j].Order
		}
		if orderA != orderB {
			return orderA < orderB
		}

		if finalRules[i].AgentDepth != finalRules[j].AgentDepth {
			return finalRules[i].AgentDepth < finalRules[j].AgentDepth
		}

		return finalRules[i].Path < finalRules[j].Path
	})

	// Print rules (body only, without front matter)
	for _, rule := range finalRules {
		body, err := extractBody(rule.Path)
		if err != nil {
			return fmt.Errorf("failed to read rule file %s: %w", rule.Path, err)
		}
		fmt.Println(body)
	}

	fmt.Printf("* * *\n\nEnd of additional context for %s. Continue.\n", filePath)
	return nil
}

// extractBody extracts the body (content after front matter) from a file
func extractBody(filePath string) (string, error) {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return "", err
	}

	contentStr := string(content)
	if len(contentStr) < 8 || contentStr[0:3] != "---" {
		return contentStr, nil
	}

	// Find the closing ---
	endIdx := -1
	for i := 3; i < len(contentStr)-3; i++ {
		if contentStr[i:i+3] == "---" && (i == 3 || contentStr[i-1] == '\n') {
			endIdx = i
			break
		}
	}

	if endIdx == -1 {
		return contentStr, nil
	}

	// Return content after the closing delimiter
	body := contentStr[endIdx+3:]
	return strings.TrimLeft(body, "\n"), nil
}
